internal/profiles/koanf.go

```diff
@@ -0,0 +1,360 @@
+func (k KoanfConfig) ProfileNames() (profileNames []string) {
+	keySet := make(map[string]struct{})
+	mainKoanfKeys := k.KoanfInstance().All()
+	for key := range mainKoanfKeys {
+		// Do not add Active profile koanf key to profileNames
+		if key == options.RootActiveProfileOption.KoanfKey {
+			continue
+		}
+
+		pName := strings.Split(key, ".")[0]
+		if _, ok := keySet[pName]; !ok {
+			keySet[pName] = struct{}{}
+			profileNames = append(profileNames, pName)
+		}
+	}
+
+	return profileNames
+}
+
+// The profile name must contain only alphanumeric characters, underscores, and dashes
+// The profile name cannot be empty
+func (k KoanfConfig) ValidateProfileNameFormat(pName string) (err error) {
+	if pName == "" {
+		return fmt.Errorf("invalid profile name: profile name cannot be empty")
+	}
+
+	re := regexp.MustCompile(`^[a-zA-Z0-9\_\-]+$`)
+	if !re.MatchString(pName) {
+		return fmt.Errorf("invalid profile name: '%s'. name must contain only alphanumeric characters, underscores, and dashes", pName)
+	}
+
+	return nil
+}
+
+func (k KoanfConfig) ChangeActiveProfile(pName string) (err error) {
+	if err = k.ValidateExistingProfileName(pName); err != nil {
+		return err
+	}
+
+	err = k.KoanfInstance().Set(options.RootActiveProfileOption.KoanfKey, pName)
+	if err != nil {
+		return fmt.Errorf("error setting active profile: %w", err)
+	}
+
+	return nil
+}
+
+// The profile name must exist
+func (k KoanfConfig) ValidateExistingProfileName(pName string) (err error) {
+	if pName == "" {
+		return fmt.Errorf("invalid profile name: profile name cannot be empty")
+	}
+
+	pNames := k.ProfileNames()
+
+	if !slices.ContainsFunc(pNames, func(n string) bool {
+		return n == pName
+	}) {
+		return fmt.Errorf("invalid profile name: '%s' profile does not exist", pName)
+	}
+
+	return nil
+}
+
+// The profile name format must be valid
+// The new profile name must be unique
+func (k KoanfConfig) ValidateNewProfileName(pName string) (err error) {
+	if err = k.ValidateProfileNameFormat(pName); err != nil {
+		return err
+	}
+
+	pNames := k.ProfileNames()
+
+	if slices.ContainsFunc(pNames, func(n string) bool {
+		return n == pName
+	}) {
+		return fmt.Errorf("invalid profile name: '%s'. profile already exists", pName)
+	}
+
+	return nil
+}
+
+func (k KoanfConfig) GetProfileKoanf(pName string) (subKoanf *koanf.Koanf, err error) {
+	if err = k.ValidateExistingProfileName(pName); err != nil {
+		return nil, err
+	}
+
+	// Create a new koanf instance for the profile
+	subKoanf = koanf.New(".")
+	err = subKoanf.Load(confmap.Provider(k.KoanfInstance().Cut(pName).All(), "."), nil)
+	if err != nil {
+		return nil, fmt.Errorf("error marshalling koanf: %w", err)
+	}
+
+	return subKoanf, nil
+}
+
+func closeFile(f *os.File) {
+	if err := f.Close(); err != nil {
+		fmt.Printf("error closing file (%s): %v", f.Name(), err)
+	}
+}
+
+func (k KoanfConfig) WriteFile() (err error) {
+	encodedConfig, err := k.KoanfInstance().Marshal(yaml.Parser())
```

Comment:
For this function `WriteFile`, prefer https://pkg.go.dev/os#Create (which truncates if already exists) and https://pkg.go.dev/gopkg.in/yaml.v3#NewEncoder to write/overwrite yaml to file and encode via `encoder.Encode(k.All())`